src/main/java/com/infobip/bitbucket/JiraVersionGeneratorHook.java

```diff
@@ -0,0 +1,127 @@
+/**
+ * # Copyright 2016 Infobip
+ * #
+ * # Licensed under the Apache License, Version 2.0 (the "License");
+ * # you may not use this file except in compliance with the License.
+ * # You may obtain a copy of the License at
+ * #
+ * # http://www.apache.org/licenses/LICENSE-2.0
+ * #
+ * # Unless required by applicable law or agreed to in writing, software
+ * # distributed under the License is distributed on an "AS IS" BASIS,
+ * # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * # See the License for the specific language governing permissions and
+ * # limitations under the License.
+ */
+package com.infobip.bitbucket;
+
+import com.atlassian.bitbucket.commit.*;
+import com.atlassian.bitbucket.hook.repository.AsyncPostReceiveRepositoryHook;
+import com.atlassian.bitbucket.hook.repository.RepositoryHookContext;
+import com.atlassian.bitbucket.repository.RefChange;
+import com.atlassian.bitbucket.repository.Repository;
+import com.atlassian.bitbucket.setting.*;
+import com.google.common.collect.ImmutableList;
+import com.infobip.infrastructure.ClockFactory;
+import com.infobip.jira.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Iterator;
+
+public class JiraVersionGeneratorHook implements AsyncPostReceiveRepositoryHook, RepositorySettingsValidator {
+
+    private static final Logger logger = LoggerFactory.getLogger(JiraVersionGeneratorHook.class);
+
+    private final CommitService historyService;
+    private final JiraService jiraService;
+    private final ImmutableList<RepositorySettingsValidator> settingsValidators;
+
+    public JiraVersionGeneratorHook(CommitService historyService, JiraService jiraService) {
+
+        this.historyService = historyService;
+        this.jiraService = jiraService;
+
+        settingsValidators = ImmutableList.of(new ProjectKeyValidator(), new VersionPatternValidator());
+    }
+
+    @Override
+    public void postReceive(RepositoryHookContext repositoryHookContext, Collection<RefChange> refChanges) {
```

Comment:
this deserves at least one sub-level of helper methods... one english sentence (method that is, in camel case) per step?